tools/wptserve/wptserve/config.py

```diff
@@ -57,14 +104,34 @@ class Config(Mapping):
         },
         "aliases": []
     }
+    default_config_cls = Config
+
+    # Configuration properties that are computed. Each corresponds to a method
+    # _get_foo, which is called with the current data dictionary. The properties
+    # are computed in the order specified in the list.
+    computed_properties = ["log_level",
+                           "paths",
+                           "server_host",
+                           "ports",
+                           "domains",
+                           "not_domains",
+                           "all_domains",
+                           "domains_set",
+                           "not_domains_set",
+                           "all_domains_set",
+                           "ssl_config"]
 
     def __init__(self,
                  logger=None,
                  subdomains=set(),
                  not_subdomains=set(),
+                 config_cls=None,
                  **kwargs):
 
-        self.log_level = kwargs.get("log_level", "DEBUG")
+        self._data = self._default.copy()
```

Comment:
Nit: perhaps we can make `self._default` a `frozendict` for an extra layer of safety.

Reply:
Not sure it's worth implementing this or importing a third party package just for this.